ch_pipeline/processing/base.py

```diff
@@ -464,6 +545,55 @@ def queue_job(script, submit=True):
         os.system(cmd % fh.name)
 
 
+def classify_failed(
+    dir: Union[Path, str], tags: list, patterns: dict = {}
+) -> Dict[str, list]:
+    """Analyze the cause of crashed jobs.
+
+    Parameters
+    ----------
+    dir : pathlib.Path or str
+        directory to find tags
+    tags : list
+        tags to check
+    patterns : Dict[str, list]
+        dictionary of patterns to check for. Keys are expected
+        categories and values are lists of regex patterns to
+        check for.
+
+    Returns
+    -------
+    crashes : Dict[str, list]
+        tags associated with each pattern. Any unmatched tags are
+        added with the key "other"
+    """
+    failed = {k: [] for k in list(patterns.keys())}
+
+    for tag in tags:
+        file = Path(dir) / tag / "job" / "jobout.log"
+        if not file.is_file():
+            continue
+        # Get the end of the file, 50 lines is probably enough
+        # to find errors but it could be increased
+        process = sp.Popen(
+            ["tail", "-n", "300", str(file)],
+            stdout=sp.PIPE,
+            shell=False,
+            universal_newlines=True,
+        )
+        stdout, _ = process.communicate()
+        # Clean up the input
+        stdout = stdout.encode("unicode-escape").decode()
```

Comment:
I don't think there's a real need to use a subprocess for this. You can do it all in Python.

Reading the last N lines is slightly tricky to do efficiently, but given that there's no good reason to use exactly N lines what I would do is read the last N*M bytes where M is your estimate of the average length of a line in bytes (maybe ~60 characters). So open the file, seek to the starting position with `f.seek(-N * M, os.SEEK_END)` and then `f.read()`.